### ISSUE TITLE
Make this turn-key

### ISSUE THREAD

**Issue by Lindenk** [invoiceninja/dockerfiles#79]
Currently the docker image does not host the application and also doesn't run several of its required services periodically. Because of this, setup takes much more time and effort, creates many more points of failure, and requires more manual effort in porting to pretty much any deployment system than it should.

Docker is all about having self contained applications, I propose that invoiceninja should be one of them.

**Comment by lalop** [comment]
Hello @Lindenk, I'm not sure to understand your point, do you mean the image should contains it's own database and server application ?

**Comment by Lindenk** [comment]
Most php apps on docker hub include whatever they need to actually host a webapp, such as a webserver if there isn't a backend as part of the application, or any scripts that need to run periodically. Since this is a php application, the webserver needs to be set up and configured with php support and other custom options. This should be up to the docker image creators, not people trying to use the application. Of course custom scripts should also be handled within the image because this is functionality that's only relevant to the developers and shouldn't be an extra step for the whoever is using it.

Databases are usually handled in their own containers and connected to over a docker network. This is usually because someone might want to set to custom storage options such as clustering and whatnot for their persistent data.

The bottom line is it shouldn't matter what language or framework was used to create the application. It should just work without needing to manually interact with the underlying setup.

**Comment by welkie** [comment]
@Lindenk 
The way it's set up right now in the [example docker-compose.yml file](https://github.com/invoiceninja/dockerfiles/blob/master/docker-compose/docker-compose.yml) is with three containers: app, web, and cron. Since you agree that using a separate container for the database is good practice, are you proposing that those three containers representing the app be merged into one?

**Comment by Lindenk** [comment]
Yes, exactly! 

The current docker compose file expects a custom nginx config specific to invoice ninja so it logically makes  sense to bundle the custom web hosting setup as part of the image. Also, requiring a separate container to run just for cron is just another step everyone using invoice ninja will need to port to whatever deployment system they're using. Since it is a (rather critical) part of invoice ninja, it would make sense for it to run within the same container as the rest of the application. Usually environment variables are used for configuration, for instance time between running artisan commands in this case.

**Comment by welkie** [comment]
@Lindenk I agree with that. From what I've seen before with PHP apps dockerized, they provide an Apache version and an Nginx version if they have the time to support both. Or, they just choose one and that is how the dockerized version of the app is hosted.

**Comment by lalop** [comment]
@Lindenk unfortunately I don't have enought time to work on an apache version but contributer are really welcome :)
For the cron part, a docker image should be seen has a process executor, since the cron's command can't be run by the web server process it makes sense, for me at least, to manage them in an other container. Can you give me some exemple of official docker image of php application or all is packaged in one container ?

**Comment by schemen** [comment]
I for one are very against everything in one container, it basically goes against what docker is for: One process per container.

What could be done is that a custom nginx container is build from you guys (invoiceninja-nginx) to make it easier. That's what I've actually built here: https://github.com/schemen/docker-ttrss

**Comment by Lindenk** [comment]
@lalop Just quickly scanning the official docker images, Jenkins, Nextcloud and WordPress are a few php applications that host their own webserver and any other processes that are needed by the program in the same image. In fact, I don't think any official images expect the user to manually setup php specific stuff to run a container. I can make the dockerfile and whatnot. To not break backwards compatibility, it would probably be a good idea to give it a new tag.

@schemen I disagree. Docker is about one application per container. It just so happens that it's easy to spread core functionality into multiple files in php. Pretty much every other backend has support for coroutines/threads and hosts its own webserver so they can be one process. And if you count threading, over half of docker hub wouldn't follow that rule anyway.

**Comment by welkie** [comment]
@schemen The only issue with multiple layers of Docker images is discoverability. How would people know that there is an image available that consolidates the app and a web server?

**Comment by schemen** [comment]
@Lindenk It's best practice: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#dont-install-unnecessary-packages

@welkie With a Readme and a docker-compose file provided in the project I suppose.

**Comment by Lindenk** [comment]
@schemen Exactly. The best practice is: 
```
Each container should have only one concern. Decoupling applications into multiple containers makes it easier to scale horizontally and reuse containers. For instance, a web application stack might consist of three separate containers, each with its own unique image, to manage the web application, database, and an in-memory cache in a decoupled manner.

Limiting each container to one process is a good rule of thumb, but it is not a hard and fast rule. For example, not only can containers be spawned with an init process, some programs might spawn additional processes of their own accord...
```

The proposal is to put the few processes needed to run the web application inside the same container. If several custom images would need to be made and linked together for the sole purpose of correctly running one specific web application, doesn't it make more sense to keep them in the same container instead?

**Comment by schemen** [comment]
@Lindenk PHP & Nginx are different components, it's about containing a task (serving http, resolving php, executing cron). You could argue that you could run it with apache I suppose but that would reduce scalability. You would also be missing those php cron jobs. 
Those containers are not much linked as much as on the same network. With docker-compose 3+ you don't have linking anymore really, it also is not recommended and deprecated.

It would communicate over normal networking, as those applications do anyway. The stack would look like that:

```
version: '3'
services
  app:
    image: invoiceninja/invoiceninja
    networks:
      - internal

  web:
    image: invoiceninja/invoiceninja-nginx
    networks:
      - internal
      - public

networks:
  internal:
```
^ This is what I have running currently (of course with all the other stuff like reverse proxy configured and persistent storage but this is just to show)


If you want to keep the lightness of nginx and php-fpm + the scalabilty that comes with it, you should absolutely never combine different applications in one container. That is the reason why for example the gitlab-omnibus container image is horrible.

**Comment by Lindenk** [comment]
@schemen They aren't different components when you have to specifically modify one for the sole purpose of working with the other. The modified nginx or Apache is not useful for any other purpose other than running with invoice ninja. There will never be a case were this specific nginx is used with any other application and there will never be a case where any more or less than one instance of this nginx is ever used with one instance of invoice ninja. The same can be said for the cron job. The existence of either of these extra processes is only a quark of php anyway. Any other backend would have both webhosting and periodic processes built in. It doesn't make sense to separate processes that are both specific to and also dependent on one another. 

Also yes I know how docker networks work, I'm just using link as shorthand for "have them talk to each other over a network".

**Comment by lalop** [comment]
@Lindenk I know the wordpress and nextcloud image pretty well since I used their work as a base for the invoiceninja's image.
In this 2 exemple, an apache version is provided, this solve the webserver container and we can provide an apache version here as I do for the alpine version.
For cron task, wordpress does them with http request, so this is out of the topic and nextcloud as a good issue  on the subject https://github.com/nextcloud/docker/issues/134 the conclusion is to keep cron in a separate container.

**Comment by Lindenk** [comment]
Alright, that's fair. I think the argument is to have invoiceninja host itself more than keeping a cron task running, since advanced configuration might be required in some scenarios.

I can contribute a pull request for adding a webserver if you'd like.

**Comment by lalop** [comment]
@Lindenk for sure with pleasure, do you mean an apache version of the image ?

**Comment by Lindenk** [comment]
@lalop Yeah, or nginx. Whichever you'd prefer

**Comment by lalop** [comment]
I prefer to keep a phpfpm version and an apache other, no nginx.

**Comment by anojht** [comment]
I am not sure if this is beneficial to the maintainer. But I have a [fork ](https://github.com/anojht/invoiceninja)of this repo where I have integrated both Nginx and the Cron job into the container. The only thing I left out was the DB since in my view having a separate DB instance is just more beneficial than integrating one. It is ready to run provided you have a db instance already deployed. 

I have been running this container without issues for numerous version of invoiceninja with upgrades and all working absolutely fine.

While the container is near fool proof, it comes configured assuming you will be using HTTPS with a valid SSL cert. Otherwise you will need to modify the Nginx config. I will not amend this because I don't believe in running a payment system without proper security methods in place.

I would be more than happy to help the maintainer, @lalop, integrate the changes I have made in my [repo](https://github.com/anojht/invoiceninja) to achieve a better self contained app service.

**Comment by lalop** [comment]
@anojht thank you to point to this image but as discussed here we are not looking to maintain an allinone image here.

@Lindenk will you do a pr with the apache version ? Let me know if need some help on it.

**Comment by Lindenk** [comment]
Ok, I'll put together a PR, I've just been a little busy recently.

**Comment by lalop** [comment]
don’t worry

**Comment by Mzngit** [comment]
So there isn't much of a delay between when Hilel posts the newest version on invoiceninja.org and when @lalop creates a new docker image, but there is often a several day delay. I'm a newbie to docker but is there a way to change dockerfile so that when its run it downloads Hilel's latest version?
This probably doesn't make sense, but would it help removing the version number from the following dockerfile line? 
RUN curl -o ninja.zip -SL https://download.invoiceninja.com/ninja-v${INVOICENINJA_VERSION}.zip \
I know that https://download.invoiceninja.com/ automatically downloads the latest version of the. ZIP file.

Alternatively, are there any instructions out there on perhaps the building of the docker image ourselves, or perhaps having a script that builds it with the latest version ?
Thanks.

**Comment by lalop** [comment]
@Mzngit I'm working to make this image official to have automatic build when new version are released, Downloading the last version when the container is created is possible but this means the image doesn't contain the application and we can't tag the image like the application.

**Comment by Technetium1** [comment]
Can you have the best of both worlds [with stack deploy](https://docs.docker.com/engine/swarm/stack-deploy/)?

**Comment by codedge** [comment]
There are now [guides how to run IN with different webserver](https://github.com/invoiceninja/dockerfiles/wiki) (in separate containers) and also one guide with having the web server on the host system and only use IN in a docker image.

This should provide enough configuration options to run IN with docker. It is not planned to release a single image with everything bundled inside.
